File: 0078.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = [[]]
        for num in nums:
            for i in range(len(res)):
                res.append(res[i] + [num])
        return res


class Solution:
    def helper(self, nums, res, pos):
        for i in range(len(res)):
            res.append(res[i] + [nums[pos]])
        if pos == len(nums) - 1:
            return None
        return self.helper(nums, res, pos + 1)

    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = [[]]
        self.helper(nums, res, 0)
        return res


class Solution:
    def dfs(self, nums, index, path, res):
        res.append(path)
        for i in range(index, len(nums)):
            self.dfs(nums, i + 1, path + [nums[i]], res)
        return res

    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []
        self.dfs(nums, 0, [], res)
        return res


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        res = []
        for i in range(2**n, 2**(n + 1)):
            bitmask = bin(i)[3:]
            res.append([nums[j] for j in range(n) if bitmask[j] == '1'])
        return res

File: 0078.cpp
// First time
/*
    backtracking
    T: O(2^n)/S: O(2^n)
*/
// e.g. [1,2,3]: [], [1], [12], [123], [13], [2], [23], [3]
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ret;
        vector<int> sub;
        dfs(nums, ret, sub, 0);
        return ret;
    }
    void dfs(vector<int> &nums, vector<vector<int>> &ret, vector<int> &sub, int i) {
        ret.push_back(sub);
        for (int j = i; j < nums.size(); ++j) {
            sub.push_back(nums[j]);
            dfs(nums, ret, sub, j + 1);
            sub.pop_back();
        }
    }
};

// e.g. [1,2,3]: [123], [12], [13], [1], [23], [2], [3], []
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ret;
        vector<int> sub;
        dfs(nums, ret, sub, 0);
        return ret;
    }
    void dfs(vector<int> &nums, vector<vector<int>> &ret, vector<int> &sub, int i) {
        if (i == nums.size()) {
            ret.push_back(sub);
            return;
        }
        // choose nums[i]
        sub.push_back(nums[i]);
        dfs(nums, ret, sub, i + 1);
        // not choose nums[i]
        sub.pop_back();
        dfs(nums, ret, sub, i + 1);
    }
};

/*
    []:        []
    [1]:       [], [1]
    [1, 2]:    [], [1], [2], [1, 2]
    [1, 2, 3]: [], [1], [2], [1, 2], [3], [1, 3], [2, 3], [1, 2, 3]
    The new elements will append at all the previous subsets

    There are 2^n number of subset
    T: O(2^n)/S: O(1)
*/
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ret = {{}};
        for (int i = 0; i < nums.size(); ++i) {
            int n = ret.size();
            for (int j = 0; j < n; ++j) {
                auto tmp = ret[j];
                tmp.push_back(nums[i]);
                ret.push_back(tmp);
            }
        }
        return ret;
    }
};